nemo/collections/asr/models/hybrid_rnnt_ctc_bpe_models.py

```diff
@@ -0,0 +1,185 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+from typing import Dict, List, Optional, Union
+
+from omegaconf import DictConfig, ListConfig, OmegaConf, open_dict
+from pytorch_lightning import Trainer
+
+from nemo.collections.asr.metrics.wer_bpe import WERBPE, CTCBPEDecoding, CTCBPEDecodingConfig
+from nemo.collections.asr.models.rnnt_models import EncDecRNNTBPEModel
+from nemo.utils import logging, model_utils
+
+
+class EncDecHybridRNNTCTCBPEModel(EncDecRNNTBPEModel):
+    """Base class for encoder decoder RNNT-based models with auxiliary CTC decoder/loss and subword tokenization."""
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+        # Convert to Hydra 1.0 compatible DictConfig
+        super().__init__(cfg=cfg, trainer=trainer)
+        cfg = model_utils.convert_model_config_to_dict_config(cfg)
+        cfg = model_utils.maybe_update_config_version(cfg)
+
+        # setup auxiliary CTC decoder if needed
+        if 'aux_ctc' in cfg:
+            with open_dict(cfg):
+                if self.tokenizer_type == "agg":
+                    cfg.aux_ctc.decoder.vocabulary = ListConfig(vocabulary)
+                else:
+                    cfg.aux_ctc.decoder.vocabulary = ListConfig(list(vocabulary.keys()))
+
+            if cfg.aux_ctc.decoder["num_classes"] < 1:
+                logging.info(
+                    "\nReplacing placholder number of classes ({}) with actual number of classes - {}".format(
+                        cfg.aux_ctc.decoder["num_classes"], len(vocabulary)
+                    )
+                )
+                cfg.aux_ctc.decoder["num_classes"] = len(vocabulary)
+
+            # Setup CTC decoding
+            ctc_decoding_cfg = self.cfg.aux_ctc.get('decoding', None)
+            if ctc_decoding_cfg is None:
+                ctc_decoding_cfg = OmegaConf.structured(CTCBPEDecodingConfig)
+                with open_dict(self.cfg.ctc_decoding):
+                    self.cfg.aux_ctc.decoding = ctc_decoding_cfg
+            self.ctc_decoding = CTCBPEDecoding(self.cfg.ctc_decoding.decoder, tokenizer=self.tokenizer)
+
+            # Setup CTC WER
+            self.ctc_wer = WERBPE(
+                decoding=self.ctc_decoding,
+                use_cer=self.cfg.ctc_decoding.get('use_cer', False),
+                dist_sync_on_step=True,
+                log_prediction=self.cfg.get("log_prediction", False),
+            )
+
+    def change_vocabulary(
+        self,
+        new_tokenizer_dir: Union[str, DictConfig],
+        new_tokenizer_type: str,
+        decoding_cfg: Optional[DictConfig] = None,
```

Comment:
should we make this rnnt_decoding_cfg? Or keep it as decoding_cfg ? Im ok with either.

Reply:
It is better to keep it the same as the RNNT so that it makes it easier to extend it.

Reply:
Cool